File: app/src/main/java/xyz/xfqlittlefan/easytest/data/QuestionSet.java
package xyz.xfqlittlefan.easytest.data;

import static xyz.xfqlittlefan.easytest.util.UtilClass.QUESTION_BANK_ID;
import static xyz.xfqlittlefan.easytest.util.UtilClass.QUESTION_SET_ID;

import androidx.annotation.Keep;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

@Keep
public class QuestionSet {
    private String name;
    private String id;
    private String url;
    private List<Set> set;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public void setUrl(String url) {
        this.url = url;
    }

    public List<Set> getSet() {
        return set;
    }

    public void init() {
        for (int i = 0; i < set.size(); i++) {
            set.get(i).savedQuestionSet = this;
            set.get(i).init();
        }
    }

    @Keep
    public static class Set {
        private QuestionSet savedQuestionSet;
        private Set savedSet;
        private String name;
        private String description;
        private String url;
        private String id;
        private final Boolean random = false;
        private List<Set> children;

        public String getName() {
            return name;
        }

        public String getDescription() {
            return description;
        }

        public String getUrl() {
            return url;
        }

        public String getId() {
            return id;
        }

        public void setId(String id) {
            this.id = id;
        }

        public Boolean getRandom() {
            return random;
        }

        public List<Set> getChildren() {
            return children;
        }

        public void init() {
            if (children != null) {
                for (int i = 0; i < children.size(); i++) {
                    children.get(i).savedSet = this;
                    children.get(i).savedQuestionSet = savedQuestionSet;
                    children.get(i).init();
                }
            }
        }

        public QuestionSet getQuestionSet() {
            if (savedQuestionSet != null) {
                return savedQuestionSet;
            } else {
                return savedSet.getQuestionSet();
            }
        }

        public List<Set> getList() {
            if (savedSet != null) {
                return savedSet.children;
            } else if (savedQuestionSet != null) {
                return savedQuestionSet.set;
            } else {
                return new ArrayList<>();
            }
        }

        public String getQuestionSetUrl() {
            if (getQuestionSet() == null) {
                return "";
            }
            return getQuestionSet().url;
        }

        public Integer getIndex() {
            int index = 114514;
            List<Set> set = getList();
            for (int i = 0; i < set.size(); i++) {
                if (set.get(i).equals(this)) {
                    index = i;
                    break;
                }
            }
            return index;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;
            Set set = (Set) o;
            return Objects.equals(name, set.name) && Objects.equals(description, set.description) && Objects.equals(url, set.url) && Objects.equals(id, set.id) && Objects.equals(random, set.random);
        }

        @Override
        public int hashCode() {
            return Objects.hash(name, description, url, id, random);
        }
    }
}

File: app/src/main/java/xyz/xfqlittlefan/easytest/data/QuestionSet.kt
package xyz.xfqlittlefan.easytest.data

import androidx.annotation.Keep
import java.util.*

@Keep
class QuestionSet {
    lateinit var name: String
    lateinit var id: String
    lateinit var url: String
    lateinit var set: List<Set>

    fun init() {
        if (this::set.isInitialized) {
            set.forEach {
                it.savedQuestionSet = this
                it.init()
            }
        }
    }

    @Keep
    class Set {
        lateinit var savedQuestionSet: QuestionSet
        private lateinit var savedSet: Set
        lateinit var name: String
        lateinit var description: String
        lateinit var url: String
        lateinit var id: String
        var random = false
        lateinit var children: List<Set>
        val isChildrenInitialized get() = this::children.isInitialized

        val questionSet: QuestionSet?
            get() = when {
                this::savedQuestionSet.isInitialized -> savedQuestionSet
                this::savedSet.isInitialized -> savedSet.questionSet
                else -> null
            }
        private val list: List<Set>
            get() = when {
                this::savedSet.isInitialized -> savedSet.children
                this::savedQuestionSet.isInitialized -> savedQuestionSet.set
                else -> listOf()
            }
        val questionSetUrl: String
            get() = questionSet?.url ?: ""
        val index: Int
            get() {
                var index = 114514
                val set = list
                for (i in set.indices) {
                    if (set[i] == this) {
                        index = i
                        break
                    }
                }
                return index
            }

        fun init() {
            if (this::children.isInitialized) {
                children.forEach {
                    if (this::savedQuestionSet.isInitialized) it.savedQuestionSet = savedQuestionSet
                    it.savedSet = this
                    it.init()
                }
            }
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other == null || javaClass != other.javaClass) return false
            val set = other as Set
            return name == set.name && description == set.description && url == set.url && id == set.id && random == set.random
        }

        override fun hashCode(): Int {
            return Objects.hash(name, description, url, id, random)
        }
    }
}